docs/contributing/999-rfcs/2024-03-24-snapshot-tests.md

```diff
@@ -0,0 +1,73 @@
+---
+title: "#5958 Cloud Snapshot Tests"
+description: Spec and design for adding cloud snapshots to the Wing test framework
+---
+
+# Cloud Snapshot Tests
+
+- **Author(s)**: @eladb
+- **Submission Date**: 2024-03-24
+- **Stage**: Proposal
+- **Stage Date**: 2024-03-24
+
+## Requirements
+
+We would like to be able to capture a snapshot of the synthesized output when compiling to cloud
+targets, and use it to protect against regressions in CI, without having to deploy to the cloud.
+
+For example, let's say I am creating a Wing library with a resource that supports both `sim` and
+`tf-aws`. I am writing a bunch of tests for it. 
+
+I can run these tests in the simulator (`wing test`) and can also run them on AWS through `wing test
+-t tf-aws`.
+
+As my code **or its dependencies** are updated, I'd like to continuously run all these tests to make
+sure my library is not broken. Running the tests against `sim` is easy (I can just run `wing test`
+in my CI build environment), but it's not realistic to run all these tests against the real cloud
+target, given it could take a very long time and could also be very expensive.
+
+In some scenarios (such as the Wing SDK tests), users will prefer to setup a continuous integration
+workflow and run tests in the cloud, but for the most part running all my tests on the cloud for
+every commit is too expensive and requires a complicated setup (i.e. an AWS account), so I'd like to
+use a more lightweight approach...
+
+> NOTE: This is not https://github.com/winglang/wing/issues/1214 (which is about using snapshots for
+> assertions).
```

Comment:
Can you explain how the use case / problems solved by #1214 is different than the use case / problems addressed in the RFC? It seems to me like they're overlapping - maybe you can add some detail how they're different (both read to me like they're mechanisms that can be used to take snapshots of IaC configuration to make a test succeed or fail).

Reply:
I'll add some details, but they are simply different use cases. #1214 is about adding a general mechanism to write snapshot-based assertions in tests that developers author (e.g. I want to assert that response returned from a service hasn't changed). This RFC is about asserting that the Wing compiler outputs (IAC configs) haven't regressed after they've been tested in the cloud.

But I can see how this could be confusing and I am wondering if maybe we can either give the snapshots this RFC refers to a different name (any suggestions?) or change the CLI switch names to `--cloud-snapshots` or something like that, so that when we introduce #1214 we can leverage the `--snapshot` switch for them (because that would be more familiar).

Maybe we can call these "synthshots"?

Reply:
I'll keep the term "snapshots" for now.